File: src/integrations/orca.ts
import { spawnCli } from '../workers/spawn-utils.js';

/**
 * Orca workspace awareness for post-run cleanup.
 *
 * When a run's --repo is an Orca-managed worktree (e.g. ~/orca/workspaces/<name>),
 * plain `git worktree remove` would leave a ghost card in the Orca app. These
 * helpers detect that case and remove through orca-cli instead.
 *
 * Everything here is best-effort and gracefully degrades: missing binary,
 * app not running, malformed output -> "not an Orca workspace" / no-op.
 */

export interface OrcaWorktreeRef {
  /** Full Orca worktree id: `<repoId>::<worktreePath>` */
  id: string;
  path: string;
}

type CliRunner = (
  cmd: string,
  args: string[],
  opts: { cwd: string; timeoutMs: number },
) => Promise<{ exitCode: number; stdout: string; timedOut: boolean }>;

const defaultRunner: CliRunner = (cmd, args, opts) => spawnCli(cmd, args, opts);

/** Extract the JSON object from CLI output that may carry non-JSON noise lines. */
function extractJson(text: string): unknown {
  const start = text.indexOf('{');
  if (start < 0) return null;
  try {
    return JSON.parse(text.slice(start)) as unknown;
  } catch {
    return null;
  }
}

/** Pure matcher over parsed `orca worktree ps` output: exact path match (slash-normalized). */
export function matchOrcaWorktree(psOutput: unknown, repoPath: string): string | null {
  if (psOutput === null || typeof psOutput !== 'object') return null;
  const result = (psOutput as { result?: unknown }).result;
  if (result === null || typeof result !== 'object') return null;
  const worktrees = (result as { worktrees?: unknown }).worktrees;
  if (!Array.isArray(worktrees)) return null;
  const normalize = (p: string) => p.replace(/\/+$/, '');
  const target = normalize(repoPath);
  for (const wt of worktrees) {
    if (wt === null || typeof wt !== 'object') continue;
    const { id, path } = wt as { id?: unknown; path?: unknown };
    if (typeof id === 'string' && typeof path === 'string' && normalize(path) === target) {
      return id;
    }
  }
  return null;
}

/** Resolve the Orca worktree id for repoPath, or null when it is not Orca-managed. */
export async function findOrcaWorktreeByPath(
  repoPath: string,
  runner: CliRunner = defaultRunner,
): Promise<string | null> {
  try {
    const r = await runner('orca', ['worktree', 'ps', '--json'], { cwd: repoPath, timeoutMs: 15_000 });
    if (r.timedOut || r.exitCode !== 0) return null;
    return matchOrcaWorktree(extractJson(r.stdout), repoPath);
  } catch {
    return null;
  }
}

/** Remove an Orca-managed workspace (card + directory). Returns true on success. */
export async function dropOrcaWorkspace(
  worktreeId: string,
  cwd: string,
  runner: CliRunner = defaultRunner,
): Promise<boolean> {
  try {
    const r = await runner(
      'orca',
      ['worktree', 'rm', '--worktree', `id:${worktreeId}`, '--force'],
      { cwd, timeoutMs: 30_000 },
    );
    return !r.timedOut && r.exitCode === 0;
  } catch {
    return false;
  }
}

/** Best-effort Orca repo registration: `orca repo add --path <repoPath> --json`. */
export async function ensureOrcaRepo(
  repoPath: string,
  runner: CliRunner = defaultRunner,
): Promise<boolean> {
  try {
    const r = await runner('orca', ['repo', 'add', '--path', repoPath, '--json'], { cwd: repoPath, timeoutMs: 15_000 });
    return !r.timedOut && r.exitCode === 0;
  } catch {
    return false;
  }
}

/** Create an Orca worktree for a worker slot. Returns worktree path or null on any failure. */
export async function createOrcaWorktree(
  repoPath: string,
  name: string,
  runner: CliRunner = defaultRunner,
): Promise<string | null> {
  try {
    const r = await runner(
      'orca',
      ['worktree', 'create', '--name', name, '--repo', `path:${repoPath}`, '--json'],
      { cwd: repoPath, timeoutMs: 30_000 },
    );
    if (r.timedOut || r.exitCode !== 0) return null;
    const json = extractJson(r.stdout) as { result?: { path?: string; worktree?: { path?: string } } } | null;
    return json?.result?.path ?? json?.result?.worktree?.path ?? null;
  } catch {
    return null;
  }
}

/** List Orca worktrees that belong to the given repo (path prefix match). Best-effort. */
export async function listOrcaWorktrees(
  repoPath: string,
  runner: CliRunner = defaultRunner,
): Promise<string[]> {
  try {
    const r = await runner('orca', ['worktree', 'ps', '--json'], { cwd: repoPath, timeoutMs: 15_000 });
    if (r.timedOut || r.exitCode !== 0) return [];
    const parsed = extractJson(r.stdout) as { result?: { worktrees?: Array<{ path?: string }> } } | null;
    const wts = parsed?.result?.worktrees;
    if (!Array.isArray(wts)) return [];
    return wts.map((w) => w.path).filter((p): p is string => typeof p === 'string' && p.startsWith(repoPath));
  } catch {
    return [];
  }
}

File: internal/integrations/orca.go
// Orca workspace awareness for post-run cleanup: the Go port of
// src/integrations/orca.ts. When a run's --repo is an Orca-managed worktree
// (e.g. ~/orca/workspaces/<name>), plain `git worktree remove` would leave
// a ghost card in the Orca app. These helpers detect that case and remove
// through orca-cli instead. Everything here is best-effort and gracefully
// degrades: missing binary, app not running, malformed output -> "not an
// Orca workspace" / no-op.
package integrations

import (
	"encoding/json"
	"strings"

	"github.com/FreePeak/devagent/internal/spawn"
)

// OrcaWorktreeRef mirrors OrcaWorktreeRef.
type OrcaWorktreeRef struct {
	// ID is the full Orca worktree id: `<repoId>::<worktreePath>`.
	ID string
	// Path is the worktree directory.
	Path string
}

// OrcaRunner is the CLI seam (mirrors the TS CliRunner injection);
// production uses spawn.RunCli.
type OrcaRunner func(cmd string, args []string, opts spawn.Options) spawn.Result

func defaultOrcaRunner(cmd string, args []string, opts spawn.Options) spawn.Result {
	return spawn.RunCli(cmd, args, opts)
}

// extractJSON mirrors extractJson: pull the first `{`-anchored JSON object
// out of CLI output that may carry non-JSON noise lines. ok=false when no
// object parses.
func extractJSON(text string) (json.RawMessage, bool) {
	start := strings.Index(text, "{")
	if start < 0 {
		return nil, false
	}
	trimmed := text[start:]
	// json.Unmarshal over the full tail fails on trailing noise, so probe
	// successive decodes the way TS JSON.parse(text.slice(start)) would
	// fail: find the longest prefix that parses as one object.
	dec := json.NewDecoder(strings.NewReader(trimmed))
	var out json.RawMessage
	if err := dec.Decode(&out); err != nil {
		return nil, false
	}
	return out, true
}

// MatchOrcaWorktree mirrors matchOrcaWorktree: pure matcher over parsed
// `orca worktree ps` output; exact path match with trailing slashes
// normalized. Returns "" when there is no match (TS null).
func MatchOrcaWorktree(psOutput any, repoPath string) string {
	if psOutput == nil {
		return ""
	}
	obj, ok := psOutput.(map[string]any)
	if !ok {
		return ""
	}
	result, ok := obj["result"].(map[string]any)
	if !ok {
		return ""
	}
	worktrees, ok := result["worktrees"].([]any)
	if !ok {
		return ""
	}
	normalize := func(p string) string { return strings.TrimRight(p, "/") }
	target := normalize(repoPath)
	for _, wt := range worktrees {
		m, ok := wt.(map[string]any)
		if !ok {
			continue
		}
		id, idOK := m["id"].(string)
		path, pathOK := m["path"].(string)
		if idOK && pathOK && normalize(path) == target {
			return id
		}
	}
	return ""
}

// FindOrcaWorktreeByPath mirrors findOrcaWorktreeByPath: resolve the Orca
// worktree id for repoPath, or "" when it is not Orca-managed.
func FindOrcaWorktreeByPath(repoPath string, runner OrcaRunner) string {
	if runner == nil {
		runner = defaultOrcaRunner
	}
	r := runner("orca", []string{"worktree", "ps", "--json"}, spawn.Options{Dir: repoPath, TimeoutMs: 15000})
	if r.TimedOut || r.ExitCode != 0 {
		return ""
	}
	raw, ok := extractJSON(r.Stdout)
	if !ok {
		return ""
	}
	var parsed any
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return ""
	}
	return MatchOrcaWorktree(parsed, repoPath)
}

// DropOrcaWorkspace mirrors dropOrcaWorkspace: remove an Orca-managed
// workspace (card + directory). Reports success/failure without throwing.
func DropOrcaWorkspace(worktreeID, cwd string, runner OrcaRunner) bool {
	if runner == nil {
		runner = defaultOrcaRunner
	}
	r := runner("orca",
		[]string{"worktree", "rm", "--worktree", "id:" + worktreeID, "--force"},
		spawn.Options{Dir: cwd, TimeoutMs: 30000})
	return !r.TimedOut && r.ExitCode == 0
}

// EnsureOrcaRepo mirrors ensureOrcaRepo: best-effort Orca repo registration
// (`orca repo add --path <repoPath> --json`).
func EnsureOrcaRepo(repoPath string, runner OrcaRunner) bool {
	if runner == nil {
		runner = defaultOrcaRunner
	}
	r := runner("orca", []string{"repo", "add", "--path", repoPath, "--json"},
		spawn.Options{Dir: repoPath, TimeoutMs: 15000})
	return !r.TimedOut && r.ExitCode == 0
}

// CreateOrcaWorktree mirrors createOrcaWorktree: create an Orca worktree
// for a worker slot. Returns the worktree path or "" on any failure.
func CreateOrcaWorktree(repoPath, name string, runner OrcaRunner) string {
	if runner == nil {
		runner = defaultOrcaRunner
	}
	r := runner("orca",
		[]string{"worktree", "create", "--name", name, "--repo", "path:" + repoPath, "--json"},
		spawn.Options{Dir: repoPath, TimeoutMs: 30000})
	if r.TimedOut || r.ExitCode != 0 {
		return ""
	}
	raw, ok := extractJSON(r.Stdout)
	if !ok {
		return ""
	}
	var parsed struct {
		Result struct {
			Path     string `json:"path"`
			Worktree struct {
				Path string `json:"path"`
			} `json:"worktree"`
		} `json:"result"`
	}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return ""
	}
	if parsed.Result.Path != "" {
		return parsed.Result.Path
	}
	return parsed.Result.Worktree.Path
}

// ListOrcaWorktrees mirrors listOrcaWorktrees: list Orca worktrees that
// belong to the given repo (path prefix match). Best-effort.
func ListOrcaWorktrees(repoPath string, runner OrcaRunner) []string {
	if runner == nil {
		runner = defaultOrcaRunner
	}
	r := runner("orca", []string{"worktree", "ps", "--json"}, spawn.Options{Dir: repoPath, TimeoutMs: 15000})
	if r.TimedOut || r.ExitCode != 0 {
		return []string{}
	}
	raw, ok := extractJSON(r.Stdout)
	if !ok {
		return []string{}
	}
	var parsed struct {
		Result struct {
			Worktrees []struct {
				Path string `json:"path"`
			} `json:"worktrees"`
		} `json:"result"`
	}
	out := []string{}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return out
	}
	for _, wt := range parsed.Result.Worktrees {
		if strings.HasPrefix(wt.Path, repoPath) {
			out = append(out, wt.Path)
		}
	}
	return out
}
